cardano-crypto-class/src/Cardano/Crypto/DSIGN/SchnorrSecp256k1.hs

```diff
@@ -162,10 +169,23 @@ instance DSIGNAlgorithm SchnorrSecp256k1DSIGN where
            useAsCStringLen bs $ \(bsp, sz) ->
              copyPtr skp (castPtr bsp) sz
   rawSerialiseSigDSIGN (SigSchnorrSecp256k1 sigPSB) = psbToByteString sigPSB
-  rawSerialiseVerKeyDSIGN (VerKeySchnorrSecp256k1 vkPSB) = psbToByteString vkPSB
+  {-# NOINLINE rawSerialiseVerKeyDSIGN #-}
+  rawSerialiseVerKeyDSIGN (VerKeySchnorrSecp256k1 vkPSB) = 
+    unsafeDupablePerformIO . psbUseAsSizedPtr vkPSB $ \pkbPtr -> do
+      res <- psbCreateSized $ \bsPtr -> do
+        res' <- secpXOnlyPubkeySerialize secpCtxPtr bsPtr pkbPtr
+        when (res' /= 1) (error "rawSerialiseVerKeyDSIGN: Failed to serialise.")
+      pure . psbToByteString $ res
   rawSerialiseSignKeyDSIGN (SignKeySchnorrSecp256k1 skPSB) = psbToByteString skPSB
+  {-# NOINLINE rawDeserialiseVerKeyDSIGN #-}
   rawDeserialiseVerKeyDSIGN bs = 
-    VerKeySchnorrSecp256k1 <$> psbFromByteStringCheck bs
+    unsafeDupablePerformIO . unsafeUseAsCStringLen bs $ \(ptr, _) -> do
+      let dataPtr = castPtr ptr
+      let loc = psbZero
+      res <- psbUseAsSizedPtr loc $ \outPtr -> do
```

Comment:
There are already such tests in place in `cardano-crypto-tests` and not one of them fails.

Reply:
Oh, yeah I see that `psbFromBytes` pads it with zeros. Nevertheless it is still very dangerous for the reason you pointed out @michaelpj Doing something like this:

```haskell
loc1 = psbZero :: PinnedSizedBytes 32
loc2 = psbZero :: PinnedSizedBytes 32
````

will very likely cause ghc to allocate a single ByteArray for all occurances of `PinnedSizedBytes 32`, which means the same buffer will be mutated in different calls, i.e. mutating `loc1`, will also mutate `loc2`

I think in general `psbZero` is not a good function to have in this API. We should put a huge warning on it that the buffer should never be mutated

Reply:
Would `NOINLINE`ing it solve the problem you mention @lehins?

Reply:
> Would NOINLINEing it solve the problem

@kozross Unlikely and I would certainly not rely on any example where it does work, since under slightly different conditions it might not. See an example of a very similar problem with unsafeThaw and this whole thing we are talking about is in essence an `unsafeThaw`:
https://gitlab.haskell.org/ghc/ghc/-/issues/12656#note_347996

Reply:
> Perhaps we can make sure it doesn't get shared by giving it a proxy argument or something.

@michaelpj That won't be a good solution either, cause ghc could just letfloat it and you will end up in in the same situation

GHC can transform this:
```
loc1 = psbZero (Proxy @32)
loc2 = psbZero (Proxy @32)
```
into this:
```
let commonPureBuffer = psbZero (Proxy @32)
let loc1 = commonPureBuffer
let loc2 = commonPureBuffer
```

Reply:
@kozross See this comment: https://gitlab.haskell.org/ghc/ghc/-/issues/12656#note_125499

Reply:
Oh boy, I love it when the documentation lies to you.

Reply:
Also @lehins - does this extend to any use of `psbFromBytes` in general?

Reply:
You know, I believe so. That is why it is best to use `psbCreate` and make a big disclaimer on `psbFromBytes` that the resulting buffer should not be mutated. In any case it's implementation is terribly inefficient, so I have no idea why anyone would use it.

Reply:
I would argue that it should just be deprecated in general. It's only used (supposedly) in testing code, its inefficiency is known (it even has a note saying this!), and given its additional troubles, I don't see why it should even exist.

Reply:
> I would argue that it should just be deprecated in general. 

Big :+1: from me.